taichi/program/async_utils.cpp

```diff
@@ -162,19 +175,72 @@ TaskMeta *get_task_meta(IRBank *ir_bank, const TaskLaunchRecord &t) {
     if (auto clear_list = stmt->cast<ClearListStmt>()) {
       meta.output_states.emplace(clear_list->snode, AsyncState::Type::list);
     }
-    // TODO: handle SNodeOpStmt etc.
     return false;
   });
 
-  // We are being conservative here: if there are any non-element-wise
-  // accesses (e.g., a = x[i + 1]), we don't treat it as completely
-  // overwriting the value state (e.g., for i in x: x[i] = 0).
-  for (auto &state : meta.output_states) {
-    if (state.type == AsyncState::Type::value) {
-      if (meta.element_wise.find(state.snode) == meta.element_wise.end()) {
-        meta.input_states.insert(state);
+  std::unordered_set<SNode *> kernel_forces_no_activate(
+      t.kernel->no_activate.begin(), t.kernel->no_activate.end());
+
+  std::unordered_set<SNode *> mask_state_inserted;
+  auto insert_mask_states_for_activation = [&](SNode *s) {
+    while (s) {
+      if (kernel_forces_no_activate.count(s) > 0) {
+        break;
+      }
+      if (mask_state_inserted.count(s) > 0) {
+        // already handled by other activations
+        break;
+      }
+      mask_state_inserted.insert(s);
+
+      // Do not record dense SNodes' mask states.
+      if (s->type != SNodeType::dense && s->type != SNodeType::place &&
+          s->type != SNodeType::root) {
+        meta.input_states.emplace(s, AsyncState::Type::mask);
+        meta.output_states.emplace(s, AsyncState::Type::mask);
+        if (is_gc_able(s->type)) {
+          meta.input_states.emplace(s, AsyncState::Type::allocator);
+          meta.output_states.emplace(s, AsyncState::Type::allocator);
+        }
       }
+      s = s->parent;
     }
+  };
+
+  for (auto &snode : activates) {
+    insert_mask_states_for_activation(snode);
+  }
+  for (auto &snode : deactivates) {
+    insert_mask_states_for_activation(snode);
```

Comment:
nit: `insert_mask_states_for_activation ` could be a bit misleading here, given it's also used for `deactivates`... Hard to name this one, but how about `_bottom_up`, which at least describes what it does (and also `s/insert_value_states_for_deactivation/insert_value_states_top_down`)?